**IMA-602 Herramientas básicas para el Análisis de Datos**

# Tarea Control de Versiones

**Autores:** Ernesto Pedrero Alecoy - Carolina Reyes Miro - Angel Santos Sanzana - Cristobal Zambrano Rabanal

**Profesores:** Walter Gómez y Nicolás Caro

**Fecha:** 12 de junio 2022


---



Considere el archivo de datos correspondiente a su grupo adjunto a la tarea en la plataforma. El
archivo contiene 4 columnas: Año, Mes, Fecha (como variable continua) y CO2 (concentración). El
objetivo final de la tarea es ajustar una función F(t) que describa aproximadamente la dependencia
de la concentración respecto a la fecha. Para ello se deben seguir los siguientes pasos, utilizando la
librería scipy.optimize para optimizar.

# Parte 1
El dato de CO2 debe ser un número real positivo. Identifique el intervalo de tiempo continuo
(meses consecutivos) más extenso que cumple tres condiciones:

* El intervalo termina con la última observación de Septiembre 2020.
* El intervalo comienza con una observación con un dato válido de CO2.
* En todo el intervalo más del 95% de las observaciones tienen un dato de CO2 válido.

Extraiga ese intervalo como la base de datos a ser utilizada en esta tarea.


In [1]:
import pandas as pd

# Read a comma-separated values (csv) file into DataFrame.
# La separacion esta dada por ;, por lo que hay que cambiarlo.

# Se usa el encoding latin1 para que pueda leer todo el archivo sin problemas

Datos_svc= pd.read_csv("Grupo_3_data_CO2_short.csv", sep=',', encoding = 'latin1')

Datos_svc=Datos_svc.set_axis(['Yr', 'Mn', 'Date', 'CO2'], axis=1)
Datos_svc

Yr  Mn       Date     CO2
0    1957   6  1957.4548  312.79
1    1957   7  1957.5370  -99.99
2    1957   8  1957.6219  -99.99
3    1957   9  1957.7068  -99.99
4    1957  10  1957.7890  -99.99
..    ...  ..        ...     ...
755  2020   5  2020.3716  409.48
756  2020   6  2020.4563  409.75
757  2020   7  2020.5383  410.18
758  2020   8  2020.6230  410.78
759  2020   9  2020.7077  411.09

[760 rows x 4 columns]

In [2]:
# Se muestran cuantos datos son negativos ('no válidos').
Nendataframe=pd.DataFrame(Datos_svc.T.iloc[3]<=0)
Nendataframe.value_counts()

CO2  
False    676
True      84
dtype: int64

In [3]:
# Se muestra el tipo de dato y si existen datos nulos por columna.
Datos_svc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 760 entries, 0 to 759
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Yr      760 non-null    int64  
 1   Mn      760 non-null    int64  
 2   Date    760 non-null    float64
 3   CO2     760 non-null    float64
dtypes: float64(2), int64(2)
memory usage: 23.9 KB


In [4]:
negatividad=[True]
Error=[]
for i in range(759,-1,-1):

    
    if Datos_svc.T.iloc[3][i]<=0:
        negatividad.append(True)
    else:
        negatividad.append(False)
    
    dfposi=pd.DataFrame(negatividad)

    errora=(dfposi.value_counts()[True]-1)/(dfposi.value_counts()[False]+dfposi.value_counts()[True]-1)
    Error.append(errora)

Error.reverse()
negatividad.reverse()
Error

[0.11052631578947368,
 0.11067193675889328,
 0.10949868073878628,
 0.1083223249669749,
 0.10714285714285714,
 0.10596026490066225,
 0.10477453580901856,
 0.1049136786188579,
 0.10372340425531915,
 0.10252996005326231,
 0.10266666666666667,
 0.10146862483311081,
 0.10026737967914438,
 0.10040160642570281,
 0.09919571045576407,
 0.09798657718120805,
 0.09811827956989247,
 0.09690444145356662,
 0.09568733153638814,
 0.09446693657219973,
 0.0945945945945946,
 0.09472259810554803,
 0.0948509485094851,
 0.09362279511533243,
 0.09239130434782608,
 0.09115646258503401,
 0.08991825613079019,
 0.09004092769440655,
 0.09016393442622951,
 0.08891928864569083,
 0.08904109589041095,
 0.0877914951989026,
 0.08653846153846154,
 0.08528198074277854,
 0.08402203856749312,
 0.08275862068965517,
 0.08149171270718232,
 0.08160442600276625,
 0.0817174515235457,
 0.08183079056865465,
 0.08055555555555556,
 0.07927677329624479,
 0.07799442896935933,
 0.07670850767085077,
 0.07541899441340782,
 0.0741258741258

In [5]:
bolean=0
columna=0
for i in range(0,760):
    if Error[i]<0.05 and bolean==0 and negatividad[i]==False:
        bolean=1
        columna=i

columna

97

In [6]:
# Se muestra el resultado final, donde Datos_limpios corresponde a 
# un dataFrame que cumple con todas las condiciones indicadas.
Datos_limpios=Datos_svc[columna:len(Error)]
Datos_limpios

Yr  Mn       Date     CO2
97   1965   7  1965.5370  319.39
98   1965   8  1965.6219  319.68
99   1965   9  1965.7068  -99.99
100  1965  10  1965.7890  -99.99
101  1965  11  1965.8740  320.17
..    ...  ..        ...     ...
755  2020   5  2020.3716  409.48
756  2020   6  2020.4563  409.75
757  2020   7  2020.5383  410.18
758  2020   8  2020.6230  410.78
759  2020   9  2020.7077  411.09

[663 rows x 4 columns]

In [7]:
# Se muestran cuantos datos son negativos ('no válidos').
Nendataframe=pd.DataFrame(Datos_limpios.T.iloc[3]<=0)
Nendataframe.value_counts()

CO2  
False    630
True      33
dtype: int64

# Parte 2
Cree una nueva columna ’CO2 completa’ que rellene los datos de CO2 faltantes de su base de
datos utilizando la siguiente metodología:
* Identificar cada intervalo que comienza y termina con una observación válida, pero que
tiene entremedio solo observaciones faltantes de CO2, por ejemplo como el siguiente
(mostrando solo fecha y CO2):
${. . . (1958.7068, 315.33)(1958.7890, −99.99)(1958.8740, −99.99)(1959.9562, −99.99)(1959.0411, 315.09) . . .}$,

* Rellenar los valores intermedios del intervalo utilizando la recta que une los extremos
del intervalo. En el ejemplo anterior sería la recta $C(t)=\frac{1}{0.3343}[(t-1958.7068)\cdot 315.09-(t-1959.0411)\cdot 315.33]$, que genera la siguiente secuencia rellenada:

${. . . (1958.7068, 315.33)(1958.7890, 315.271)(1958.8740, 315.210)(1959.9562, 315.151)(1959.0411, 315.09) . . .}$

In [8]:
# En esta celda se busca los intervalos pedidos, empezando con un positivo y terminando con positivo
# y conteniendo entremedio solo valores negativos
indice_neg=[]

for i in range(97,760):

    if  Datos_limpios.CO2[i]<0:
        indice_neg.append(i)

nums = sorted(set(indice_neg))
gaps = [[s, e] for s, e in zip(nums, nums[1:]) if s+1 < e]
edges = iter(nums[:1] + sum(gaps, []) + nums[-1:])

intervalos_ddd=list(zip(edges, edges))
intervalos=[]

for i in range(len(intervalos_ddd)):
    intervalos.append((intervalos_ddd[i][0]-1,intervalos_ddd[i][1]+1))

intervalos

[(98, 101),
 (108, 110),
 (114, 116),
 (116, 118),
 (127, 130),
 (132, 138),
 (139, 141),
 (149, 151),
 (161, 163),
 (169, 171),
 (171, 173),
 (173, 177),
 (250, 252),
 (272, 274),
 (294, 296),
 (319, 321),
 (402, 404),
 (503, 505),
 (512, 514),
 (576, 579),
 (585, 587),
 (593, 595),
 (724, 726),
 (732, 734)]

In [9]:
# Aqui obtengo los intervalos de solo negativos, para su facil trabajo en la siguiente celda
m=[]
n=[]

for i in range(len(intervalos)):
    m.append((Datos_limpios.CO2[intervalos[i][0]]-Datos_limpios.CO2[intervalos[i][1]])/(Datos_limpios.Date[intervalos[i][0]]-Datos_limpios.Date[intervalos[i][1]]))
    n.append(-m[i]*Datos_limpios.Date[intervalos[i][0]]+Datos_limpios.CO2[intervalos[i][0]])
    
intervalos_ddd

[(99, 100),
 (109, 109),
 (115, 115),
 (117, 117),
 (128, 129),
 (133, 137),
 (140, 140),
 (150, 150),
 (162, 162),
 (170, 170),
 (172, 172),
 (174, 176),
 (251, 251),
 (273, 273),
 (295, 295),
 (320, 320),
 (403, 403),
 (504, 504),
 (513, 513),
 (577, 578),
 (586, 586),
 (594, 594),
 (725, 725),
 (733, 733)]

In [10]:
# y datos por los que reemplazar los negativos.
y=[] # y = m*x + n

# El indice i corresponde al numero de dato que es negativo,
# y[0] corresponde al primer dato negativo, y[1] corresponde al segundp dato negativo.
for i in range(len(intervalos)):
    for j in range(intervalos_ddd[i][0],intervalos_ddd[i][1]+1):
        y.append(m[i]*Datos_limpios.Date[j]+n[i])
        
y

[319.8450178500593,
 320.00478778262595,
 320.77007181328645,
 321.03,
 320.9335930735931,
 321.5640707964602,
 321.59592920353975,
 321.6384399999997,
 321.84341399999994,
 322.04838799999925,
 322.2465860000002,
 322.4515599999995,
 322.30746287128727,
 323.8271813285455,
 325.0235547576294,
 325.47000000000116,
 325.7503349282297,
 325.56107887579265,
 325.3868147476578,
 325.21275611967303,
 333.1867684021536,
 335.93395973154384,
 338.7750000000001,
 341.8181208053684,
 352.39500000000044,
 365.40992818671475,
 366.5255033557047,
 376.96078571428734,
 377.39539285714454,
 377.89843301435394,
 379.37096947935356,
 403.83100478468896,
 405.4920825852787]

In [11]:
# Se crea una nueva columna (una copia de la columna CO2) 
# para reemplazar los datos con la interpolación.
Datos_limpios["CO2_Completa"]=Datos_limpios.CO2
Datos_limpios

C:\Users\Angel\AppData\Local\Temp\ipykernel_7340\155735336.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Datos_limpios["CO2_Completa"]=Datos_limpios.CO2


Yr  Mn       Date     CO2  CO2_Completa
97   1965   7  1965.5370  319.39        319.39
98   1965   8  1965.6219  319.68        319.68
99   1965   9  1965.7068  -99.99        -99.99
100  1965  10  1965.7890  -99.99        -99.99
101  1965  11  1965.8740  320.17        320.17
..    ...  ..        ...     ...           ...
755  2020   5  2020.3716  409.48        409.48
756  2020   6  2020.4563  409.75        409.75
757  2020   7  2020.5383  410.18        410.18
758  2020   8  2020.6230  410.78        410.78
759  2020   9  2020.7077  411.09        411.09

[663 rows x 5 columns]

In [12]:
contador=0
for i in range(Datos_limpios.CO2.count()):

    if Datos_limpios.iloc[i,4]<0:
        
        Datos_limpios.iloc[i,4]=y[contador]
        contador=contador+1

# dataframe con la nueva columna sin negativos a través del metodo dado.
Datos_limpios 

Yr  Mn       Date     CO2  CO2_Completa
97   1965   7  1965.5370  319.39    319.390000
98   1965   8  1965.6219  319.68    319.680000
99   1965   9  1965.7068  -99.99    319.845018
100  1965  10  1965.7890  -99.99    320.004788
101  1965  11  1965.8740  320.17    320.170000
..    ...  ..        ...     ...           ...
755  2020   5  2020.3716  409.48    409.480000
756  2020   6  2020.4563  409.75    409.750000
757  2020   7  2020.5383  410.18    410.180000
758  2020   8  2020.6230  410.78    410.780000
759  2020   9  2020.7077  411.09    411.090000

[663 rows x 5 columns]

In [13]:
# Se muestran cuantos datos son negativos ('no válidos').
Nendataframe=pd.DataFrame(Datos_limpios.T.iloc[4]<=0)
Nendataframe.value_counts()

CO2_Completa
False           663
dtype: int64

Aquí se puede observar que la cantidad de datos negativos es cero.